Add ref output to CoreCompile target and add CopyRefAssembly task

jcouv opened the pull request:
Hopefully, this should unblock the work by the msbuild team to consume ref assemblies.

I still need to verify that the dependencies for the new task gets packaged properly into the nuget package.

@tmat @jaredpar @dotnet/roslyn-compiler

jcouv commented:
@dotnet-bot test windows_debug_vs-integration_prtest please

jcouv commented:
@jaredpar @agocke @dotnet/roslyn-compiler for second review please.

jaredpar commented on src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs:
```diff
@@ -0,0 +1,84 @@
+            }
+
+            var source = ExtractMvid(SourcePath);
+            var destination = ExtractMvid(DestinationPath);
```
Need to consider the case where ExtractMvid fails.  Consider that the developer corrupts a file on disk, it's held with a non-sharing lock, or a simple text file is accidentally put into the location.  This version of the code would cause the CopyAssemblyRef task to fail.  

I think the design should be that we keep the old one as an optimization.  Any error trying to implement that optimization should fall back to the non-optimized Copy path (with a log message).

jaredpar commented on src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs:
```diff
@@ -0,0 +1,84 @@
+                return false;
+            }
+
+            if (!File.Exists(DestinationPath))
```
Making us redundant to extractMvid failures would also allow you to remove this check.  Or at least centralize it in ExtractMvid.

jcouv commented:
@jaredpar I addressed your feedback.

Here's the current behavior for CopyRefAssembly:
1. no source: error
2. no destination: copy (silent)
3. bad source: copy (with High importance message)
4. bad destination: copy (with Low importance message)
5. source and destination match (happy case): skip (with Low importance message)
6. source and destination differ: copy (silent)
7. copy fails: error (with details)

FYI @KirillOsenkov

KirillOsenkov commented:
FYI @andygerlicher @jeffkl @rainersigwald @cdmihai

KirillOsenkov commented:
FYI @ljcollins25

ljcollins25 commented:
Just a thought. Is it necessary to interpret the contents of the file? Can it just check the bytes?

jcouv commented:
@ljcollins25 We could do a full contents comparison, but we already have a hash built-into the assembly. Seems cheaper to just check a couple bytes.

jcouv commented:
@dotnet-bot test windows_release_vs-integration_prtest  please

jaredpar commented on src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs:
```diff
@@ -0,0 +1,95 @@
+    /// If the source and destination ref assemblies are different,
+    /// this task copies the source over to the destination.
+    /// </summary>
+    public class CopyRefAssembly : Task
```
Consider `sealed`

jaredpar commented on src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs:
```diff
@@ -0,0 +1,95 @@
+                return false;
+            }
+
+            if (File.Exists(DestinationPath))
```
Consider documenting the strategy that we discussed.  Essentially the no-copy is an optimization and in the case of other failures we should fall back to the default which is a simple copy.

jcouv commented:
I produced an official build from this feature branch in order to unblock the msbuild folks. It contains updated bits for csc.exe, Csc task, and CoreCompile target, as well as the new CopyRefAssembly task.

https://dotnet.myget.org/feed/roslyn/package/nuget/Microsoft.CodeAnalysis.Build.Tasks/2.3.0-refout-61528-05

In parallel, I will continue to optimize what goes into ref assemblies (for instance stripping private methods out)

FYI @ljcollins25 @KirillOsenkov @andygerlicher @jeffkl @rainersigwald @cdmihai
